### ISSUE TITLE
Inconsistent "Internal Failure" on function resources during sam deploy

### ISSUE THREAD

**Issue by AbeGellis** [aws/aws-sam-cli#2876]
### Description:
I am converting a series of unconnected Lambda functions to a shared SAM codebase (migrating to new function names, so not overwriting anything outside of that codebase). I have successfully added, built, and deployed several functions, but several of them are now causing updates to fail on attempts to deploy largely unaltered new versions. Specifically, I'm seeing an UPDATE_FALURE event for the AWS::Lambda::Version resource with a ResourceStatusReason of "Internal Failure". Moreover, these failures are inconsistent - different deployment attempts result in different sets of function versions hitting Internal Failure. 

I have an unusual setup to begin with. All function resources have AutoPublishAlias configured. Additionally, all function resources have an environment variable set to a deployment-unique parameter value (usually automatically set to the deployment timestamp). This is because I was encountering the issue described in https://github.com/aws/aws-sam-cli/issues/2872 and, combined with AutoPublishAlias, found it to be obstructing my deployments (unaltered code would be uploaded, but Lambda would reject a new version with unchanged code). I was able to successfully upload once with this strategy.

I am able to build and locally invoke the function resources that fail to update. They use the same role and layer configuration - a single psycopg2-binary layer referenced by ARN, not included in the template but in my account. Other updating functions that do not seem to produce Internal Failure have two things in common:

- A reference to another layer, local to the project, in addition to the psycopg2-binary layer
- I was able to successfully upload them at least once before the deployment-unique parameter workaround

### Steps to reproduce:
This began to happen after I added the deployment-unique parameter and may be connected to it, but beyond that I have no idea what would be the minimal configuration necessary to reproduce this. The debug flag doesn't seem to add any useful information to the frustratingly opaque "Internal Failure" but I'm happy to try different configurations. --force-upload didn't seem to change anything.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

OS: Windows 10 Home version 2004
sam --version: 1.23.0
AWS region: us-east-1

**Comment by aahung** [comment]
Can you provide us a minimal template so that we can reproduce it? Thanks

**Comment by AbeGellis** [comment]
I've attempted to reduce this to a reproducible template without much luck, including a setup with varying parameter-driven environment variables. It consistently occurs on my actual project (as in, every time I try to deploy, not consistent in terms of which resources fail) but the opaqueness of "Internal Failure" isn't much to go off of.

Is there anything I can do, besides calling deploy with the --debug parameter, to get more information on what's happening under the hood?

**Comment by AbeGellis** [comment]
Many of the functions resources that were failing had a lot of duplicated code between them from before I had built an internal layer to manage that code. Additionally, some of them were missing external layer references. I did a pass to clean those issues up and deploy worked fine again.

No idea what the exact condition is to trigger the Internal Errors I was seeing before - some of those functions still _should_ have worked, even with duplicated legacy code in them. I'll keep an eye out as I continue to expand the template and close this issue for now.

**Comment by sigpwned** [comment]
I have hit a similar issue a couple of times, including just now. To the best of my ability to determine the underlying issue, it looks like the issue was caused by updating a tag value on a lambda function during a SAM deploy. I arrived at this conclusion by examining CloudTrail outputs and eliminating changes to my stack until the changes went through. I am making this comment less in the hopes of generating a fix, since I still don't really feel confident that I understand the issue and don't have a smoking gun, and more in the hopes that it will help someone else in the future, possibly me!
